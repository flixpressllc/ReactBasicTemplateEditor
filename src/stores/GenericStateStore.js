import { clone, traverseObject, toType } from 'happy-helpers';

/*
Allowed states will look something like this:

  allowedStates = {
    templateType: 'string',
    templateId: 'number',
    caughtErrors: 'array'
  }

  Starting values will look something like this:

    allowedStates = {
      templateType: '',
      caughtErrors: []
    }
*/

class GenericStateStore {
  constructor (allowedStates, startingValues) {
    this._setupAllowedStates(allowedStates);
    this._setupStartingValues(startingValues);
  }

  _setupAllowedStates (allowedStates) {
    if (toType(allowedStates) !== 'object') {
      throw new Error(`A generic state store cannot be instanciated without an 'allowedStates' object. ${allowedStates} given.`)
    }
    this.allowedStates = allowedStates;
  }

  _setupStartingValues (startingValues) {
    if (startingValues) {
      this.storedState = startingValues;
    }
  }

  getState (stringOrArrayOfKeys) {
    if (stringOrArrayOfKeys === undefined) return clone(this.storedState);
    if (toType(stringOrArrayOfKeys) === 'array') {
      return this._getStateByArrayOfKeys(stringOrArrayOfKeys);
    }
    if (toType(stringOrArrayOfKeys) === 'string') {
      return this._getStateByKey(stringOrArrayOfKeys);
    }
  }

  // Returns a value only
  _getStateByKey (key) {
    if (this.allowedStates[key] === undefined) {
      throw new Error(`The desired state property "${key}" from StateStore is not allowed.`);
    }
    if (['object', 'array'].indexOf(this.allowedStates[key]) > -1) {
      return clone(this.storedState[key]);
    }
    return this.storedState[key];
  }

  // Returns an object of key: value pairs
  _getStateByArrayOfKeys (arrayOfKeys) {
    let objects = arrayOfKeys.map(key => {
      return {[key]: this.getStateByKey(key)} // ES2015 computed property names
    });
    return Object.assign(...objects);
  }

  setState (obj) {
    this._storedStatesAreAllowed(obj);
    let newState = Object.assign({}, this.storedState, obj);
    if (!this._storedStatesAreEqual(newState)) {
      this.storedState = newState;
      return true;
    }
    return false;
  }

  _storedStatesAreAllowed (obj) {
    traverseObject(obj, (k, v) => {
      if (this.allowedStates[k] === undefined) {
        throw new Error(`There is no entry for "${k}" in allowed states.`);
      }
      if (this.allowedStates[k] !== toType(v)) {
        throw new Error(`Expected the type of ${k} to be ${this.allowedStates[k]}. ${toType(v)} was given.`)
      }
    })
  }

  _storedStateExists (key) {
    return this.storedState[key] !== undefined;
  }

  _storedStateValueEquals (key, newValue) {
    if (!this._storedStateExists(key)) return false;
    let stateValue = this.storedState[key];
    let theType = toType(stateValue);

    if (theType !== toType(newValue)) return false;
    if (['array','object'].indexOf(theType) > -1) {
      return JSON.stringify(newValue) === JSON.stringify(stateValue);
    }
    return newValue === stateValue;
  }

  _storedStatesAreEqual (newState) {
    let equal = true;
    traverseObject(newState, (k, v) => {
      if (!this._storedStateValueEquals(k, v)) {
        equal = false;
      }
    })
    return equal;
  }
}

export default GenericStateStore;
