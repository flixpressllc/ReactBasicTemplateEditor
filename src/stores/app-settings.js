export const YOU_TUBE_API_KEY = 'AIzaSyBxurdcfDBk6hKm3l_vk3MWHGWle47RdMQ';
export const XML_CONTAINER_ID = 'RndTemplate_HF';
export const IMAGES_CONTAINER_ID = 'CroppedImageFilenames';

const LOCAL_DEV_THUMBNAIL_PATH = '/images/';
const PRODUCTION_THUMBNAIL_PATH = '/templates/TB/';

function thumbnailPath () {
  if (window.location.hostname === 'localhost') return LOCAL_DEV_THUMBNAIL_PATH;
  return PRODUCTION_THUMBNAIL_PATH;
}

export const THUMBNAIL_URL_PREFIX = thumbnailPath();
