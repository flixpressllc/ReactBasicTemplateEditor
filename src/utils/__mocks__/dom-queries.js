'use strict';
// export function find(selector) {
//   return $(selector);
// }

var __mockElements = {};

export function getElementById (id) {
  return __mockElements[id];
}

export function __setMockElement(id, object) {
  __mockElements[id] = object;
}
