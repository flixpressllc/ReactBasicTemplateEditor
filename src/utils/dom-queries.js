export function find(selector) {
  return $(selector).get();
}

export function getElementById (id) {
  return document.getElementById(id);
}
