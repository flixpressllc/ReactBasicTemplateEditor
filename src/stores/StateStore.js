import { EventEmitter } from 'events';
import dispatcher from '../actions/dispatcher';
import { clone, traverseObject, toType } from 'happy-helpers';
import { isDevelopment } from './app-settings';

const ALLOWED_STATES = {
  templateType: 'string',
  templateId: 'number',
  caughtErrors: 'array',
  username: 'string',
  usernames: 'string'
}

class StateStore extends EventEmitter {
  constructor () {
    super();
    this.state = {
      caughtErrors: []
    };
  }

  getState (stringOrArrayOfKeys) {
    if (stringOrArrayOfKeys === undefined) return clone(this.state);
    if (toType(stringOrArrayOfKeys) === 'array') {
      return this.getStateByArrayOfKeys(stringOrArrayOfKeys);
    }
    if (toType(stringOrArrayOfKeys) === 'string') {
      return this.getStateByKey(stringOrArrayOfKeys);
    }
  }

  // Returns a value only
  getStateByKey (key) {
    if (ALLOWED_STATES[key] === undefined) {
      throw new Error(`The desired state property "${key}" from StateStore is not allowed.`);
    }
    if (['object', 'array'].indexOf(ALLOWED_STATES[key]) > -1) {
      return clone(this.state[key]);
    }
    return this.state[key];
  }

  // Returns an object of key: value pairs
  getStateByArrayOfKeys (arrayOfKeys) {
    let objects = arrayOfKeys.map(key => {
      return {[key]: this.getStateByKey(key)} // ES2015 computed property names
    });
    return Object.assign(...objects);
  }

  setState (obj) {
    this.statesAreAllowed(obj);
    let newState = Object.assign({}, this.state, obj);
    if (!this.statesAreEqual(newState)) {
      this.state = newState;
      this.emit('STATE_UPDATED');
    }
  }

  statesAreAllowed (obj) {
    traverseObject(obj, (k, v) => {
      if (ALLOWED_STATES[k] === undefined) {
        throw new Error(`There is no entry for "${k}" in StateStore's allowed states.`);
      }
      if (ALLOWED_STATES[k] !== toType(v)) {
        throw new Error(`StateStore was expecting the type of ${k} to be ${ALLOWED_STATES[k]}. ${toType(v)} was given.`)
      }
    })
  }

  stateExists (key) {
    return this.state[key] !== undefined;
  }

  stateValueEquals (key, newValue) {
    if (!this.stateExists(key)) return false;
    let stateValue = this.state[key];
    let theType = toType(stateValue);

    if (theType !== toType(newValue)) return false;
    if (['array','object'].indexOf(theType) > -1) {
      return JSON.stringify(newValue) === JSON.stringify(stateValue);
    }
    return newValue === stateValue;
  }

  statesAreEqual (newState) {
    let equal = true;
    traverseObject(newState, (k, v) => {
      if (!this.stateValueEquals(k, v)) {
        equal = false;
      }
    })
    return equal;
  }

  handleActions(action) {
    switch(action.type) {
      case 'SET_STATE':
        this.setState(action.state);
      break;
      default: break;
    }
  }
}

const stateStore = new StateStore();
dispatcher.register(stateStore.handleActions.bind(stateStore))

if (isDevelopment()) {
  window.StateStore = stateStore;
}

export default stateStore;
