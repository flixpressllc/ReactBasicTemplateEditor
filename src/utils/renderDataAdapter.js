import {xmlStringToObject, objectToXml} from './xmlAdapter';
import { XML_CONTAINER_ID, IMAGES_CONTAINER_ID,
  TOP_LEVEL_NAME_IMAGES, TOP_LEVEL_NAME_TEXT_ONLY} from '../stores/app-settings';
import { getElementById } from './dom-queries';
import { clone, convertPropKeysForJs, convertPropKeysForAsp,
  nestedPropertyTest, isObject, isNotEmpty, isEmpty,
  traverseObject, wrapObjectWithProperty, forceArray, nestedPropertyExists } from 'happy-helpers';

let startingPoint = {
  ResolutionId: 0,
  IsPreview: false
};

let startingSpecs = {
  Specs: {
    $name: 'Specs',
    $val: '',
    SpCx: {
      CSp: []
    }
  }
};

let startingSlides = {
  Slides: {
    FSlide: {
      Captions: {},
      Images: {}
    }
  }
};

let startingAudioInfo = {
  AudioInfo: {
    Name: null,
    Length: '0',
    AudioType: 'NoAudio',
    Id: '0',
    AudioUrl: null
  }
};

function getEmptyRenderedData () {
  if (isImageTemplate()) {
    return Object.assign({}, startingSlides, startingAudioInfo)
  } else {
    return Object.assign({}, startingSpecs, startingAudioInfo)
  }
}

function getEmptyStartingPoint () {
  let orderStartingPoint = clone(startingPoint);
  orderStartingPoint.RenderedData = getEmptyRenderedData();
  return orderStartingPoint;
}

function xmlContainerDiv () {
  return getElementById(XML_CONTAINER_ID);
}

function getXmlContainerValue () {
  return xmlContainerDiv().value;
}

function setXmlContainerValue (value) {
  xmlContainerDiv().value = value;
}

var getLoadedXmlAsString = function () {
  return getXmlContainerValue();
};

var getLoadedXmlAsObject = function () {
  return xmlStringToObject(getLoadedXmlAsString());
};

function isImageTemplate () {
  return getTopLevelXmlName() === TOP_LEVEL_NAME_IMAGES;
}

var getTopLevelXmlName = function () {
  const allowedNames = [TOP_LEVEL_NAME_TEXT_ONLY, TOP_LEVEL_NAME_IMAGES]
  let topLevelName = '';
  traverseObject(getLoadedXmlAsObject(), key => {
    if ( allowedNames.indexOf(key) !== -1 )
    topLevelName = key;
  });
  if (topLevelName === '') throw new Error('no TopLevelName discovered in ' + getLoadedXmlAsString());
  return topLevelName;
};


var convertSpecsToReactData = function (givenXmlObj) {
  if (!nestedPropertyExists(givenXmlObj,'RenderedData.Specs.SpCx.CSp')) {
    return [];
  }
  let specs = forceArray(clone(givenXmlObj.RenderedData.Specs.SpCx.CSp));

  let nameValuePairs = [];
  for (var i = 0; i < specs.length; i++) {
    var currentFieldsArray = forceArray(specs[i].SpCx.Sp);
    var name = '';
    var value = '';
    for ( var j = 0; currentFieldsArray.length > j; j++ ) {
      name = currentFieldsArray[j].$name;
      value = currentFieldsArray[j].$val;
      let o = {};
      o[name] = value;
      nameValuePairs.push(o);
    }
  }

  return nameValuePairs;
};

function getStartingResolutionsObject (obj) {
  if (!nestedPropertyTest(obj, 'ResolutionOptions.ListItemViewModel', isNotEmpty)) {
    throw new Error('No resolutions available');
  }

  let givenResolutions = forceArray(clone(obj.ResolutionOptions.ListItemViewModel));

  // Eventual refactor for arrays of Objects?
  let resolutions = []
  for (let i=0; i < givenResolutions.length; i++) {
    resolutions.push(convertPropKeysForJs(givenResolutions[i]));
  }

  return {resolutions: resolutions, resolutionId: resolutions[0].id};
}

function getStartingAudioObject (obj) {
  if (nestedPropertyTest(obj, 'RenderedData.AudioInfo', isObject)) {
    return {audioInfo: convertPropKeysForJs(obj.RenderedData.AudioInfo)};
  }
  return {};
}

function getImagesFromHiddenField () {
  let container = getElementById(IMAGES_CONTAINER_ID);
  if (!container) return [];

  let imagesString = getElementById(IMAGES_CONTAINER_ID).value;
  let imagesArray = imagesString.split('|');
  return imagesArray.filter( val => isNotEmpty(val) );
}

function getImagesFromUnusedRenderData (obj) {
  if (nestedPropertyTest(obj, 'RenderedData.UnusedImageUrls.String', isNotEmpty)) {
    return forceArray(obj.RenderedData.UnusedImageUrls.String);
  } else {
    return [];
  }
}

function getImageBankObject (obj) {
  if (!isImageTemplate()) return {};
  let allImages = getImagesFromUnusedRenderData(obj)
    .concat(getImagesFromHiddenField())
    .reduce((a, file) => {
      if (a.indexOf(file) === -1) a.push(file);
      return a;
    }, []);

  return { imageBank: allImages };
}

function getNameValuePairsForMainCaptionFields (givenXmlObj) {
  let nameValuePairs = [];
  if (nestedPropertyTest(givenXmlObj,'RenderedData.Captions.CaptionField', isNotEmpty)) {
    let rDataCaptionFields = forceArray(givenXmlObj.RenderedData.Captions.CaptionField);

    rDataCaptionFields.map((captionField) => {
      let o = {};
      o[captionField.Label] = captionField.Value;
      nameValuePairs.push(o);
    });
  }

  return nameValuePairs;
}

function returnNameValuePairForSingleImageContainer (givenXmlObj) {
  if (nestedPropertyTest(givenXmlObj,'RenderedData.Slides.FSlide.Images.CaptionedImage', isNotEmpty)) {
    let captionedImages = forceArray(givenXmlObj.RenderedData.Slides.FSlide.Images.CaptionedImage);
    let mainImageData = [];
    captionedImages.map( (capImage, i) => {
      let capFields = forceArray(capImage.Captions.CaptionField);
      if (isEmpty(capFields)) {
        mainImageData.push({id: i, file: capImage.Filename});
      } else {
        mainImageData.push({id: i, file: capImage.Filename, captions: capFields.map(field => {
            return field.Value;
          })
        });
      }
    });
    return {ImageContainer: mainImageData};
  }
  return {};
}

function convertCaptionsToReactData (givenXmlObj) {
  let nameValuePairs = getNameValuePairsForMainCaptionFields(givenXmlObj);
  let imageContainerPair = returnNameValuePairForSingleImageContainer(givenXmlObj);

  if (isNotEmpty(imageContainerPair)) nameValuePairs.push(imageContainerPair);

  if (isEmpty(nameValuePairs)) {
    return [];
  } else {
    return nameValuePairs;
  }
}

function getNameValuePairsObj(givenXmlObj) {
  let nameValuePairsArr = isImageTemplate() ? convertCaptionsToReactData(givenXmlObj) : convertSpecsToReactData(givenXmlObj);

  return Object.assign({}, ...nameValuePairsArr);
}

function getReactStartingData () {
  let xmlObj = getLoadedXmlAsObject()[getTopLevelXmlName()];

  return [
    Object.assign(
      getImageBankObject(xmlObj),
      getStartingResolutionsObject(xmlObj),
      getStartingAudioObject(xmlObj),
      {isPreview: false},
    ),
    getNameValuePairsObj(xmlObj)
  ];

}

function orderObjectToXml (object) {
  const FIND_ORDER_REQ_TAG = /<(OrderReq[A-z]+)[^>]*>/;
  const firstString = '<?xml version="1.0" encoding="utf-16"?>\n'
  const typeDeclaration = 'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"';

  let xmlString = objectToXml(object);
  xmlString = xmlString.replace(FIND_ORDER_REQ_TAG, (match, capture) => {
    debugger;
    return `<${capture} ${typeDeclaration}>`;
  })
  return firstString + xmlString;
}

function addResolutionToOrderObj (orderObj, reactObj) {
  if (reactObj.resolutionId === undefined || reactObj.resolutionId === 0) {
    throw new Error('No ResolutionId was present');
  }
  if (reactObj.resolutionOptions === undefined) {
    throw new Error('Resolution Options are extraneous... but I want the data to match exactly. No Resolution Options were present.')
  }
  let newOrderObj = clone(orderObj);
  newOrderObj.ResolutionId = reactObj.resolutionId;
  return newOrderObj;
}

function addResolutionOptionsToOrderObj (orderObj, reactObj) {
  if (reactObj.resolutionOptions === undefined) {
    throw new Error('Resolution Options are extraneous... but I want the data to match exactly. No Resolution Options were present.')
  }
  let newOrderObj = clone(orderObj);
  newOrderObj.ResolutionOptions = reactObj.resolutionOptions.reduce((a, resObj) => {
    // TODO: test that this is actually necessary... just flipping the order
    if (isImageTemplate()){
      a.ListItemViewModel.push({
        'Id': resObj.id,
        'Name': resObj.name
      });
    } else {
      a.ListItemViewModel.push({
        'Name': resObj.name,
        'Id': resObj.id
      });
    }
    return a;
  }, {'ListItemViewModel':[]});
  return newOrderObj;
}

function sortObjectByArray (keysInOrder, obj) {
  let newObj = keysInOrder.reduce((a, key) => {
    if (!obj.hasOwnProperty(key)) return a;
    a[key] = clone(obj[key]);
    return a;
  }, {});

  traverseObject(obj, (key, value) => {
    if (newObj.hasOwnProperty(key)) return;
    newObj[key] = clone(value);
  });

  return newObj;
}

function addAudioToOrderObj (orderObj, reactObj) {
  if (reactObj.audioInfo === undefined) {
    throw new Error('No audioInfo was present');
  }
  let newOrderObj = clone(orderObj);

  let sortOrder = ['length', 'id', 'audioType', 'name', 'audioUrl']

  // TODO: test if this flip reordering is actually necessary
  if (isImageTemplate()) {
    sortOrder = ['audioUrl', 'name', 'audioType', 'length', 'id'];
  }

  let audioObject = sortObjectByArray(sortOrder, reactObj.audioInfo);

  // copy audio
  newOrderObj.RenderedData.AudioInfo = convertPropKeysForAsp(audioObject);
  return newOrderObj;
}

function addSpecsToOrderObj (orderObj, reactObj) {
  if (reactObj.ui === undefined) {
    throw new Error('No Specs were sent');
  }

  let newOrderObj = clone(orderObj);
  // Distribute Specs
  for (var i = 0; i < reactObj.ui.length; i++) {

    for (var key in reactObj.ui[i]) {
      if (reactObj.ui[i].hasOwnProperty(key)){
        var SpArray = [];

        for (var j = 0; j < reactObj.ui[i][key].length; j++) {
           SpArray.push({
            $name: reactObj.ui[i][key][j].name,
            $val: reactObj.ui[i][key][j].value
          });
        }

        newOrderObj.RenderedData.Specs.SpCx.CSp.push({
          $name: key,
          $val: 'CD|' + key + '|',
          SpCx: {
            Sp: SpArray
          }
        });

      }
    }

  }
  return newOrderObj;
}

function addImageRenderDataToOrderObject (orderObject, reactObj) {
  if (reactObj.ui === undefined) {
    throw new Error('No Image Data was sent');
  }

  let newOrderObj = clone(orderObject);
  var Captions = {CaptionField: []};
  let Images = {CaptionedImage: []};

  for (var i = 0; i < reactObj.ui.length; i++) {

    for (var key in reactObj.ui[i]) {
      if (reactObj.ui[i].hasOwnProperty(key)){

        for (var j = 0; j < reactObj.ui[i][key].length; j++) {
          if (reactObj.ui[i][key][j].type !== 'UserImageChooser') {
            Captions.CaptionField.push({
              Label: reactObj.ui[i][key][j].name,
              Value: reactObj.ui[i][key][j].value
            });
          } else {
            let chooser = reactObj.ui[i][key][j];
            chooser.value.map(imgObj => {
              Images.CaptionedImage.push({
                Captions: { CaptionField:
                  imgObj.captions.map((cap) => {
                    return {
                      Label: cap.label,
                      Value: cap.value
                    };
                  })
                },
                Filename: imgObj.file
              });
            });
          }
        }
      }
    }
  }

  newOrderObj.RenderedData.Slides.FSlide.Images = Images;

  let unusedImages = {UnusedImageUrls: { String: reactObj.imageBank}};

  let otherProps = [
    {LogoMode: 'Image'},
    {LogoId: 0},
    {LogoFilename: {}},
    {Logo3DFilename: {}}
  ]

  Object.assign(newOrderObj.RenderedData, {Captions}, unusedImages, ...otherProps);

  let sortArray = ['Slides', 'UnusedImageUrls', 'Captions', 'LogoMode', 'AudioInfo', 'LogoId', 'LogoFilename', 'Logo3DFilename'];

  newOrderObj.RenderedData = sortObjectByArray(sortArray, newOrderObj.RenderedData);

  return newOrderObj
}

function updateXmlForOrder (reactObj) {
  var orderObject = getEmptyStartingPoint();

  orderObject = addResolutionToOrderObj(orderObject, reactObj);
  orderObject = addAudioToOrderObj(orderObject, reactObj);
  orderObject.IsPreview = reactObj.isPreview;
  orderObject = addResolutionOptionsToOrderObj(orderObject, reactObj);

  if (isImageTemplate()) {
    orderObject = addImageRenderDataToOrderObject(orderObject, reactObj);
  } else {
    orderObject = addSpecsToOrderObj(orderObject, reactObj);
  }

  let sortArray = ['ResolutionId', 'IsPreview', 'ResolutionOptions', 'RenderedData'];

  if (isImageTemplate()) {
    sortArray = ['IsPreview', 'RenderedData', 'ResolutionId', 'ResolutionOptions'];
  }

  orderObject = sortObjectByArray(sortArray, orderObject);

  orderObject = wrapObjectWithProperty(orderObject, getTopLevelXmlName());
  setXmlContainerValue( orderObjectToXml(orderObject) );
}

export {
  getReactStartingData,
  updateXmlForOrder
};

export let _private = {
  getImagesFromUnusedRenderData
};
