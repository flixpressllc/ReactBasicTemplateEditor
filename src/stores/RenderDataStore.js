import { EventEmitter } from 'events';
import dispatcher from './dispatcher';
import * as Containers from '../utils/globalContainerConcerns';

import { clone, isObject } from '../utils/helper-functions';

class RenderDataStore extends EventEmitter {
  constructor () {
    super();
    this.state = {
      receivedInitial: false
    }
  }
  changeContainer(dataType, fieldName, newData) {
    let containerName = Containers.getContainerNameFor(dataType);
    Object.assign(this.containers[containerName][fieldName], newData);
    this.emit('change', {containerName, dataType, fieldName});
  }
  
  addMissingContainers() {
    if (!isObject(this.containers)) {
      this.containers = {};
    }
    Containers.getContainerNames().map(name => {
      if (!this.containers.hasOwnProperty(name)) {
        this.containers[name] = {};
      }
    })
  }

  getAll() {
    if ( ! this.state.receivedInitial) this.addMissingContainers();
    return this.containers ? clone(this.containers) : {};
  }
  
  handleActions(action) {
    switch(action.type) {
      case 'CHANGE_CONTAINER':
        this.changeContainer(action.dataType, action.fieldName, action.newData);
      break;
      case 'RECEIVE_INITIAL_CONTAINER_VALUES':
        this.state.receivedInitial = true;
        this.containers = action.containers;
        this.emit('change');
      break;
    }
  }
}

const renderDataStore = new RenderDataStore;
dispatcher.register(renderDataStore.handleActions.bind(renderDataStore))
window.renderDataStore = renderDataStore;
export default renderDataStore;
