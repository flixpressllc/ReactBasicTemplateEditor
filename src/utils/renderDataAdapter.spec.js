import * as DataAdapter from './renderDataAdapter';
import jsb from 'js-beautify';
import { XML_CONTAINER_ID, IMAGES_CONTAINER_ID } from '../stores/app-settings';
import { getSubmissionXmlFor, getMockOrderObject } from '../../specs/spec-helpers';

jest.mock('./dom-queries');

// make it pretty and make self-terminating tags consistent ( <This /> versus <This/> )
let pretty = (obj) => jsb.html(obj).replace(/(\S)\/>/g, '$1 />');

const startingTextOnlyXml = getSubmissionXmlFor('startingTextOnly');

const startingImagesXml = getSubmissionXmlFor('startingImages');

const startingImagesCroppedImagesFormFieldValue = 'HiddenFieldImage1.png|HiddenFieldImage2.jpg|HiddenFieldImage3.png|';

describe('DataAdapter', () => {
  describe('getReactStartingData', () => {
    it('returns the expected object for a new TextOnly', () => {
      require('./dom-queries').__setMockElement(XML_CONTAINER_ID, {value: startingTextOnlyXml});
      expect(DataAdapter.getReactStartingData()).toMatchSnapshot();
    });
    it('returns the expected object for a previously created TextOnly', () => {
      require('./dom-queries').__setMockElement(XML_CONTAINER_ID, {value: getSubmissionXmlFor(79)});
      expect(DataAdapter.getReactStartingData()).toMatchSnapshot();
    });

    it('returns the expected object for a new Images template', () => {
      let dom = require('./dom-queries');
      dom.__setMockElement(XML_CONTAINER_ID, {value: startingImagesXml});
      dom.__setMockElement(IMAGES_CONTAINER_ID, {value: startingImagesCroppedImagesFormFieldValue});
      expect(DataAdapter.getReactStartingData()).toMatchSnapshot();
    });
    describe('when starting with a previous render', () => {
      describe('and when images contain captions', () => {
        it('returns the expected object', () => {
          let dom = require('./dom-queries');
          dom.__setMockElement(IMAGES_CONTAINER_ID, {value: ''});
          dom.__setMockElement(XML_CONTAINER_ID, {value: getSubmissionXmlFor(92)});
          expect(DataAdapter.getReactStartingData()).toMatchSnapshot();
        });
      });
      describe('and when images do not contain captions', () => {
        it('returns the expected object', () => {
          let dom = require('./dom-queries');
          dom.__setMockElement(IMAGES_CONTAINER_ID, {value: ''});
          dom.__setMockElement(XML_CONTAINER_ID, {value: getSubmissionXmlFor(96)});
          expect(DataAdapter.getReactStartingData()).toMatchSnapshot();
        });
      });
    });

    it('does not return duplicate images from hidden field and captions in the imageBank container', () => {
      pending();

      expect();
    });
  });

  describe('updateXmlForOrder', () => {
    it('updates the container with the expected values for new TextOnly', () => {
      let dom = require('./dom-queries');
      dom.__setMockElement(XML_CONTAINER_ID, {value: startingTextOnlyXml});

      DataAdapter.updateXmlForOrder( getMockOrderObject(79));

      expect(pretty(dom.getElementById(XML_CONTAINER_ID).value)).toEqual(pretty(getSubmissionXmlFor(79)));
    });
    it('updates the container with the expected values for new Images', () => {
      let dom = require('./dom-queries');
      dom.__setMockElement(XML_CONTAINER_ID, {value: startingImagesXml});

      DataAdapter.updateXmlForOrder(getMockOrderObject(92));

      let result = pretty(dom.getElementById(XML_CONTAINER_ID).value);

      let expected = pretty(getSubmissionXmlFor(92));

      expect(result).toEqual(expected);
    });
  });
});
