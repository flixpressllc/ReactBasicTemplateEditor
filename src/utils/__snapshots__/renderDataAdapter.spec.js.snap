exports[`DataAdapter getReactStartingData returns the expected object for a new Images template 1`] = `
Object {
  "isPreview": false,
  "resolutionId": 5,
  "resolutions": Array [
    Object {
      "id": 5,
      "name": "720p",
    },
    Object {
      "id": 3,
      "name": "1080p",
    },
    Object {
      "id": 4,
      "name": "4K",
    },
  ],
  "userImages": Array [
    Object {
      "id": 0,
      "url": "DonDentonAdmin_1-23-2017_94956756.png",
    },
    Object {
      "id": 1,
      "url": "DonDentonAdmin_1-23-2017_9502787.jpg",
    },
    Object {
      "id": 2,
      "url": "DonDentonAdmin_1-23-2017_9505506.png",
    },
  ],
}
`;

exports[`DataAdapter getReactStartingData returns the expected object for a new TextOnly 1`] = `
Object {
  "isPreview": false,
  "resolutionId": 5,
  "resolutions": Array [
    Object {
      "id": 5,
      "name": "720p",
    },
    Object {
      "id": 3,
      "name": "1080p",
    },
    Object {
      "id": 4,
      "name": "4K",
    },
  ],
}
`;

exports[`DataAdapter getReactStartingData returns the expected object for a previously created TextOnly 1`] = `
Object {
  "audioInfo": Object {
    "audioType": "StockAudio",
    "audioUrl": "https://fpsound.s3.amazonaws.com/13.mp3",
    "id": 13,
    "length": 3,
    "name": "Bunny Garden",
  },
  "isPreview": true,
  "nameValuePairs": Array [
    Object {
      "name": "Icon Style",
      "value": "camera",
    },
    Object {
      "name": "Icon Character - type in only one (optional)",
      "value": "q",
    },
    Object {
      "name": "Background",
      "value": "camo",
    },
    Object {
      "name": "Icon Color",
      "value": "icon_pink",
    },
    Object {
      "name": "Main Text Color",
      "value": "main_orange",
    },
    Object {
      "name": "Subtitle Text Color",
      "value": "sub_white",
    },
    Object {
      "name": "Text Left of Icon",
      "value": "Left",
    },
    Object {
      "name": "Text Right of Icon",
      "value": "Right",
    },
    Object {
      "name": "Subtitle",
      "value": "Bottom",
    },
  ],
  "resolutionId": 5,
  "resolutions": Array [
    Object {
      "id": 5,
      "name": "720p",
    },
    Object {
      "id": 3,
      "name": "1080p",
    },
  ],
}
`;

exports[`DataAdapter updateXmlForOrder updates the container with the expected values for new TextOnly 1`] = `
"<?xml version=\"1.0\" encoding=\"utf-16\"?>
<OrderRequestOfTextOnlyRndTemplate xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\">
    <ResolutionId>5</ResolutionId>
    <RenderedData>
        <Specs name=\"Specs\" val=\"\">
            <SpCx>
                <CSp name=\"Customize Icon\" val=\"CD|Customize Icon|\">
                    <SpCx>
                        <Sp name=\"Icon Style\" val=\"camera\" />
                        <Sp name=\"Icon Character - type in only one (optional)\" val=\"q\" />
                    </SpCx>
                </CSp>
                <CSp name=\"Customize Look\" val=\"CD|Customize Look|\">
                    <SpCx>
                        <Sp name=\"Background\" val=\"camo\" />
                        <Sp name=\"Icon Color\" val=\"icon_pink\" />
                        <Sp name=\"Main Text Color\" val=\"main_orange\" />
                        <Sp name=\"Subtitle Text Color\" val=\"sub_white\" />
                    </SpCx>
                </CSp>
                <CSp name=\"Customize Text\" val=\"CD|Customize Text|\">
                    <SpCx>
                        <Sp name=\"Text Left of Icon\" val=\"Left\" />
                        <Sp name=\"Text Right of Icon\" val=\"Right\" />
                        <Sp name=\"Subtitle\" val=\"Bottom\" />
                    </SpCx>
                </CSp>
            </SpCx>
        </Specs>
        <AudioInfo>
            <AudioType>StockAudio</AudioType>
            <AudioUrl>https://fpsound.s3.amazonaws.com/13.mp3</AudioUrl>
            <Id>13</Id>
            <Length>3</Length>
            <Name>Bunny Garden</Name>
        </AudioInfo>
    </RenderedData>
    <IsPreview>true</IsPreview>
</OrderRequestOfTextOnlyRndTemplate>"
`;
